libs/sdk-core/src/ldk/store/mirroring_store.rs

```diff
@@ -0,0 +1,429 @@
+                    "UPDATE store SET value = ?1, local_version = ?2 WHERE primary_ns = ?3 AND secondary_ns = ?4 AND key = ?5",
+                    params![value, next_version, primary_ns, secondary_ns, key],
+                ).map_err(other)?;
+                next_version
+            }
+        };
+
+        tokio::task::block_in_place(|| {
+            self.handle.block_on(
+                self.remote_client
+                    .put(full_key, value.to_vec(), next_version),
+            )
+        })
+        .map_err(other)?;
+
+        conn.execute(
+            "UPDATE store SET remote_version = local_version WHERE primary_ns = ?1 AND secondary_ns = ?2 AND key = ?3",
+            params![primary_ns, secondary_ns, key],
+        ).map_err(other)?;
+
+        Ok(())
+    }
+
+    fn remove(
+        &self,
+        primary_ns: &str,
+        secondary_ns: &str,
+        key: &str,
+        _lazy: bool,
+    ) -> io::Result<()> {
+        let conn = self.pool.get().map_err(other)?;
+
+        conn.execute(
+            "DELETE FROM store WHERE primary_ns = ?1 AND secondary_ns = ?2 AND key = ?3",
+            params![primary_ns, secondary_ns, key],
+        )
+        .map_err(other)?;
+
+        let full_key = format!("{primary_ns}/{secondary_ns}/{key}");
+        tokio::task::block_in_place(|| self.handle.block_on(self.remote_client.delete(full_key)))
+            .map_err(other)?;
+
+        Ok(())
+    }
+
+    fn list(&self, primary_ns: &str, secondary_ns: &str) -> io::Result<Vec<String>> {
+        self.pool
+            .get()
+            .map_err(other)?
+            .prepare("SELECT key FROM store WHERE primary_ns = ?1 AND secondary_ns = ?2")
+            .map_err(other)?
+            .query_map(params![primary_ns, secondary_ns], |row| row.get(0))
+            .map_err(other)?
+            .collect::<Result<Vec<String>, _>>()
+            .map_err(other)
+    }
+}
+
+fn is_dirty(conn: &Connection) -> rusqlite::Result<bool> {
+    let dirty_rows: i64 = conn.query_row(
+        "SELECT count(1) FROM store WHERE local_version != remote_version",
+        [],
+        |row| row.get(0),
+    )?;
+    Ok(dirty_rows > 0)
+}
+
+async fn download<S: VersionedStore>(conn: &Connection, remote: &S) -> Result<(), Error> {
+    conn.execute("DELETE FROM store", [])?;
+
+    for (full_key, version) in remote.list().await? {
+        trace!("Downloading {full_key} @ {version} ...");
+        let parts: Vec<&str> = full_key.splitn(3, '/').collect();
+        let (primary, secondary, key) = match &parts[..] {
+            [p, s, k] => (p.to_string(), s.to_string(), k.to_string()),
+            _ => continue, // skip malformed keys
+        };
+
+        if let Some((value, version)) = remote.get(full_key).await? {
+            trace!("Got {} bytes @ {version}", value.len());
+            conn.execute(
+                "INSERT INTO store (primary_ns, secondary_ns, key, value, local_version, remote_version) VALUES (?1, ?2, ?3, ?4, ?5, ?5)",
```

Comment:
Shouldn't we only insert if remote version is larger than local? Do you have any concern of deleting all items and just copy the remote?